Tests/OpenCombineTests/PublisherTests/ZipTests.swift

```diff
@@ -0,0 +1,692 @@
+            receiveSubscription: { $0.request(.unlimited) },
+            receiveValue: { _ in
+                XCTAssertEqual(children[0].publisher.send(1), .none)
+                return Subscribers.Demand.none
+            }
+        )
+
+        zip.subscribe(downstreamSubscriber)
+
+        XCTAssertEqual(children[0].publisher.send(1), .none)
+        XCTAssertEqual(children[1].publisher.send(1), .none)
+
+        XCTAssertEqual(downstreamSubscriber.history, [.subscription("Zip"),
+                                                      .value(2)])
+    }
+
+    func testUpstreamFinishReceivedWhileSendingValue() {
+        let (children, zip) = getChildrenAndZipForArity(2)
+
+        let downstreamSubscriber = TrackingSubscriber(
+            receiveSubscription: { $0.request(.unlimited) },
+            receiveValue: { _ in
+            children[0].publisher.send(completion: .finished)
+                return Subscribers.Demand.none
+            }
+        )
+
+        zip.subscribe(downstreamSubscriber)
+
+        XCTAssertEqual(children[0].publisher.send(1), .none)
+        XCTAssertEqual(children[0].publisher.send(1), .none)
+        XCTAssertEqual(children[1].publisher.send(1), .none)
+
+        XCTAssertEqual(downstreamSubscriber.history, [.subscription("Zip"),
+                                                      .value(2)])
+    }
+
+    // NOTE about how/when Apple sends .finished on `Zip`.
+    //
+    // The documentation says:
+    //      If either upstream publisher finishes successfuly or fails with an error,
+    //      the zipped publisher does the same.
+    //
+    // This may be true for`.failed`, but it just isn't true for `.finished`.
+    // The combination of tests in here confirm Apple's actual behavior. An assesment
+    // is made to determine if it is possible for the `Zip` to send any more values
+    // downstream. Roughly speaking, if any children required to provide component
+    // values for the next value have finished, then it will be impossible for `Zip`
+    // to send any more values.
+    // This assessment is slightly complicated by the fact that `Zip` buffers _surplus_
+    // component values while waiting to complete the entire tuple.
+    //
+    // AppleRef: 002 The algorithm is currently further complicated by the fact that this
+    // assessment is not made continuously, but rather only when one of the child
+    // subcriptions sends a `.finished`. This means that Apple's behavior is inconsistent.
+    // Sometimes, the `Zip` remains alive even though no futher emissions are possible.
+    // Sometimes it finishes. Ugh.
+    //
+    // If I were in charge, `Zip` would finish as soon as it becomes impossible for it
+    // to send another value - regarless of what triggers that change in state.
+
+    func testZipCompletesOnlyAfterAllChildrenComplete() {
+        let child1Publisher = PassthroughSubject<Int, Never>()
+        let child2Publisher = PassthroughSubject<Int, Never>()
```

Comment:
A lot of tests here use `PassthroughSubject`. It'd be good if you could replace it with `CustomPublisherBase`, so that we could assert on its `CustomSubscription`'s `history`. This way we test more of the observable behavior.

Reply:
I've changed some of the tests to use `CustomPublisherBase` and `CustomSubscription`, but the rest of them also assert on the history of their downstream subscribers. Adding an upstream subscription has an impact on the downstream subscribers history as well, which I think indicates that the nature of the test changes too? Please let me know how you'd like the rest of those tests to look, or if keeping `PassthroughSubject` in them would be fine otherwise.